File: app/models/entity.rb
class Entity < ApplicationRecord
  belongs_to :contract

  has_many :fields
  has_many :actions
  has_many :relationships, inverse_of: :entity, foreign_key: :entity_id
  has_many :source_relationships, class_name: 'Relationship', inverse_of: :dependent_entity, foreign_key: :dependent_entity_id

  has_and_belongs_to_many :included_in_actions, class_name: 'Action', join_table: 'actions_entities'
end

File: app/frontend/app/models/entity.js
import Ember from 'ember';
import DS from 'ember-data';

const { computed } = Ember;

export default DS.Model.extend({

  name: DS.attr('string'),
  description: DS.attr('string'),

  contract: DS.belongsTo('contract'),
  fields: DS.hasMany('field'),
  actions: DS.hasMany('action', { inverse: 'entity' }),
  relationships: DS.hasMany('relationship', { inverse: 'entity', async: false }),
  sourceRelationships: DS.hasMany('relationship', { inverse: 'dependentEntity' }),

  includedInActions: DS.hasMany('action', { inverse: 'includedEntities' }),

  belongsToRelationships: computed.filterBy('relationships', 'cardinality', 'BELONGS_TO'),
  belongsToEntities: computed.mapBy('belongsToRelationships', 'dependentEntity'),
  hasManyRelationships: computed.filterBy('relationships', 'cardinality', 'HAS_MANY'),
  hasManyEntities: computed.mapBy('hasManyRelationships', 'dependentEntity'),

});
